src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -419,11 +399,27 @@ generateP scrutTy (pat, alt)
         require (annotOfExp alt) $ CEquals datS t
 
         (alt',sub) <- generateX alt
+        let altTy'  = annResult (annotOfExp alt')
+        altTy      <- TypeVar <$> fresh
+        altTp      <- TypeVar <$> fresh
+
+        -- Require alternative types to have the same temporality if they
+        -- do have temporalities. Otherwise defaults to TemporalityPure.
+        require (annotOfExp alt) $ CExtractTemporality altTp altTy altTy'
 
-        e' <- lift $ lift State.get
+        -- Require the alternative types without temporality to be the same.
+        require (annotOfExp alt) $ CEquals resTy altTy
+
+        -- Require return temporality to be compatible with alternative temporalities.
+        resTp' <- TypeVar <$> fresh
+        require (annotOfExp alt) $ CTemporalityJoin resTm resTp' altTp
+
+        e'    <- lift $ lift State.get
         lift $ lift $ State.put (e' { stateEnvironment = stateEnvironment e })
+        rest' <- generateP ann scrutTy resTy resTp' rest
+
+        return (((pat,alt'), sub):rest')
```

Comment:
hm. I would like to unify types `t` and `u`, where `t` has some temporality `tempT` and base type `tyT`.
If  `u` is `Temporality _ _`, then the specific temporalities must join. Otherwise, if `u` does not have temporality (e.g. just a BoolT), then default it to Pure and perform the join. If `u` is a type variable, defer the constraint.
I tried to do this with CEquals and getTemporality but that fixes the constraints too early...